### PR TITLE
Update JSON menu representation in mutations

### LINKED ISSUE #4471: Cannot assign a page to a menu item (graphql)
Cannot assign "<Page: About>": "MenuItem.category" must be a "Category" instance.

```
Traceback (most recent call last):
  File "/Users/mikail/Development/saleor-venv/lib/python3.7/site-packages/promise/promise.py", line 487, in _resolve_from_executor
    executor(resolve, reject)
  File "/Users/mikail/Development/saleor-venv/lib/python3.7/site-packages/promise/promise.py", line 754, in executor
    return resolve(f(*args, **kwargs))
  File "/Users/mikail/Development/saleor-venv/lib/python3.7/site-packages/graphql/execution/middleware.py", line 75, in make_it_promise
    return next(*args, **kwargs)
  File "/Users/mikail/Development/saleor/saleor/graphql/core/mutations.py", line 207, in mutate
    response = cls.perform_mutation(root, info, **data)
  File "/Users/mikail/Development/saleor/saleor/graphql/core/mutations.py", line 354, in perform_mutation
    instance = cls.construct_instance(instance, cleaned_input)
  File "/Users/mikail/Development/saleor/saleor/graphql/core/mutations.py", line 184, in construct_instance
    f.save_form_data(instance, data)
  File "/Users/mikail/Development/saleor-venv/lib/python3.7/site-packages/django/db/models/fields/__init__.py", line 853, in save_form_data
    setattr(instance, self.name, data)
  File "/Users/mikail/Development/saleor-venv/lib/python3.7/site-packages/django/db/models/fields/related_descriptors.py", line 211, in __set__
    self.field.remote_field.model._meta.object_name,
ValueError: Cannot assign "<Page: About>": "MenuItem.category" must be a "Category" instance.
```

### PR DESCRIPTION
**Pull request by maarcingebala**
Storefront 1.0 uses `json_content` fields to render menus in its HTML templates (to optimize rendering and not fetch the whole structure from the database every time). Those fields were however not updated when using API mutations.

This PR introduces the following changes:
- adds `update_menu` and `update_menus` calls in mutations that touch menu items
- moves the utilty functions from `saleor.dashboard.menu.utils` to `saleor.menu.utils` as they're now used outside of dashboard 1.0 as well.

Also fixes #4471

### Pull Request Checklist

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [ ] All visible strings are translated with proper context.
1. [ ] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] Database migration files are up to date.
1. [x] The changes are tested.
1. [ ] The code is documented (docstrings, project documentation).
1. [x] GraphQL schema and type definitions are up to date.
1. [x] Changes are mentioned in the changelog.

### CODE REVIEW

**Comment by NyanKiyoshi** [comment]
Maybe this PR could include a fix for #4471?

**Comment by dominik-zeglen** [comment]
Should be fixed long time ago

**Comment by NyanKiyoshi** [comment]
@dominik-zeglen the issue is not the dashboard, but the mutation not checking the user input 👍

**Comment by django-queries** [comment]
Here is the report for 0b2761bd7807e7eace05e85c21bd9dc0df1d4d81 (maarcingebala/saleor @ update-menu-json-in-api)
Base comparison is 7b3de4150086c7249a5824e212917bded235de85.



<details><summary>No differences were found. (click me)</summary>
<p>

```diff
# api.benchmark checkout
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  add billing address to checkout     	         34	         34	             20
  add shipping to checkout            	          7	          7	              0
  checkout payment charge             	         14	         14	              0
  complete checkout                   	          6	          6	              0
  create checkout                     	         41	         41	             20

# api.benchmark homepage
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  retrieve main menu                  	          5	          5	              0
  retrieve product list               	          4	          4	              0
  retrieve secondary menu             	          5	          5	              0
  retrieve shop                       	          2	          2	              0

# api.benchmark product
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  product details                     	         13	         13	              3

# api.benchmark variant
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  retrieve variant list               	          9	          9	              2
```

</p>
</details>

**Comment by maarcingebala** [comment]
@NyanKiyoshi I also fixed the issue #4471 that you've mentioned.

File: saleor/graphql/menu/bulk_mutations.py [hunk 1]
```diff
@@ -26,3 +27,9 @@ class Meta:
         description = "Deletes menu items."
         model = models.MenuItem
         permissions = ("menu.manage_menus",)
+
+    @classmethod
+    def bulk_action(cls, queryset):
+        menu_ids = set([item.menu_id for item in queryset])
```
Comment (NyanKiyoshi) [inline saleor/graphql/menu/bulk_mutations.py]: What about directly creating a set?

```suggestion
        menu_ids = {item.menu_id for item in queryset}
```

File: saleor/menu/utils.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            child_data["child_items"] = grand_children_data
+            top_item_data["child_items"].append(child_data)
+        menu_data.append(top_item_data)
+    return menu_data
+
+
+def update_menus(menus_pk):
+    menus = Menu.objects.filter(pk__in=menus_pk)
+    for menu in menus:
+        update_menu(menu)
```
Comment (NyanKiyoshi) [inline saleor/menu/utils.py]: You might want to run this inside a transaction

**Comment by maarcingebala** [comment]
@NyanKiyoshi What do you exactly mean by "updating dashboard-next"? I made fixes only in the backend.

**Comment by NyanKiyoshi** [comment]
@maarcingebala nevermind. At the time of reviewing the `name` field wasn't optional.

File: saleor/menu/utils.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        menu_item.category = linked_object
+    elif isinstance(linked_object, Collection):
+        menu_item.collection = linked_object
+    elif isinstance(linked_object, Page):
+        menu_item.page = linked_object
+
+    return menu_item.save()
+
+
+def get_menus_that_needs_update(collection=None, categories=None, page=None):
```
Comment (korycins) [inline saleor/menu/utils.py]: `get_menus_that_need_update`?

File: saleor/menu/utils.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    """Returns PrimaryKeys of Menu instances that will be affected by
```
Comment (korycins) [inline saleor/menu/utils.py]: No need to have CamelCase. `Returns primary keys of `

**Pull request merged**